core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/common/UserUniqueIDManger.java

```diff
@@ -67,20 +68,29 @@ public User addUser(String username, String profileName, AbstractUserStoreManage
     public User getUser(String uniqueId, AbstractUserStoreManager userStoreManager)
             throws UserStoreException {
 
-        String[] usernames = userStoreManager.getUserList(UserCoreClaimConstants.USER_ID_CLAIM_URI, uniqueId, null);
+        String userName = userStoreManager.getFromUserNameCache(uniqueId);
```

Comment:
Ideally this cache needs to be managed outside the user-store-manager level, as it needs to keep a reference from UUID to {tenant, user-store, user-store-local-id}

Reply:
+1. Currently, the cache is managed outside of user-store-manager.  It keeps UUID/tenantId--> to {full qualified username} 

https://github.com/wso2/carbon-kernel/blob/4.6.x/core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/common/UserIdResolverCache.java#L71